ODIN_II/SRC/HardSoftLogicMixer.cpp

```diff
@@ -0,0 +1,119 @@
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "HardSoftLogicMixer.hpp"
+
+#include <stdint.h> // INT_MAX
+#include <vector>
+
+#include "multipliers.h"       // instantiate_simple_soft_multiplier
+#include "netlist_statistic.h" // mixing_optimization_stats
+#include "odin_error.h"        // error_message
+
+HardSoftLogicMixer::HardSoftLogicMixer(const config_t& config) {
+    _allOptsDisabled = true;
+    // By default, disables all optimizations
+    parse_opt_parameters(config);
+}
+
+void HardSoftLogicMixer::parse_opt_parameters(const config_t config) {
+    if (config.mults_ratio != -1.0) {
+        MultsOpt* opt = new MultsOpt(config.mults_ratio);
+        this->_enabled_opts.push_back(opt);
+        _allOptsDisabled = false;
+    } else if (config.exact_mults != -1) {
+        MultsOpt* opt = new MultsOpt(config.exact_mults);
+        this->_enabled_opts.push_back(opt);
+        _allOptsDisabled = false;
+    }
+}
+
+void HardSoftLogicMixer::note_candidate_node(nnode_t* opNode, operation_list type) {
+    _nodes_by_opt[type].push_back(opNode);
+}
+
+bool HardSoftLogicMixer::softenable(operation_list type) {
+    bool result = false;
+    if (!this->_allOptsDisabled) {
+        for (MixingOpt* opt : _enabled_opts) {
+            if (opt->get_kind() == type)
+                return opt->enabled();
+        }
+    }
+
+    return result;
+}
+
+void HardSoftLogicMixer::map_deferred_blocks(netlist_t* netlist) {
+    if (_allOptsDisabled) {
+        return;
+    }
+    grid_statistics();
+    choose_hard_blocks(netlist);
+    soft_map_remaining_nodes(netlist);
+}
+
+void HardSoftLogicMixer::grid_statistics() {
+    if (true == _allOptsDisabled) {
+        return;
+    }
+
+    for (MixingOpt* opt : _enabled_opts) {
+        if (opt->enabled()) {
+            operation_list kind = opt->get_kind();
+            int blocks_needed = this->hard_blocks_needed(kind);
+            opt->set_blocks_needed(blocks_needed);
+        }
+    }
+}
+
+int HardSoftLogicMixer::hard_blocks_needed(operation_list opt) {
+    return _nodes_by_opt[opt].size();
+}
+
+void HardSoftLogicMixer::soft_map_remaining_nodes(netlist_t* netlist) {
+    for (unsigned int i = 0; i < operation_list::operation_list_END; i++) {
+        for (unsigned int j = 0; j < _nodes_by_opt[i].size(); j++) {
+            switch (i) {
+                case operation_list::MULTIPLY:
+                    instantiate_simple_soft_multiplier(_nodes_by_opt[i][j], PARTIAL_MAP_TRAVERSE_VALUE, netlist);
+                    break;
+                default:
+                    error_message(NETLIST, unknown_location, "%s",
+                                  "Cleanup for optimization with number: Optimization with number: %i does not have an implementation inside of HardSoftLogicMixer::soft_map_remaining_nodes\n", i);
+                    break;
+            }
+        }
+    }
+}
+
+void HardSoftLogicMixer::choose_hard_blocks(netlist_t* netlist) {
+    for (MixingOpt* opt : _enabled_opts) {
```

Comment:
the itterator will have to change to a indexed for loop once you switch to a static array

also should check if it is enabled

Reply:
Done, thanks for the comment!